### PR TITLE
[WIP] Add LMSDiscreteSchedulerTest

### PR DESCRIPTION
**Pull request by sidthekidder**
PR for #338, still to fix some tests.

### CODE REVIEW

File: tests/test_scheduler.py [hunk 1]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+    def test_timesteps(self):
+        for timesteps in [10, 50, 100, 1000]:
+            self.check_over_configs(num_train_timesteps=timesteps)
+
+    def test_betas(self):
+        for beta_start, beta_end in zip([0.0001, 0.001, 0.01, 0.1], [0.002, 0.02, 0.2, 2]):
+            self.check_over_configs(beta_start=beta_start, beta_end=beta_end)
+
+    def test_schedules(self):
+        for schedule in ["linear"]:
```
Comment (anton-l) [inline tests/test_scheduler.py:853]: We should check over all shedules that are currently supported / in use
```suggestion
        for schedule in ["linear", "scaled_linear"]:
```

File: tests/test_scheduler.py [hunk 2]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+        pass
+
+    def test_full_loop_no_noise(self):
+        scheduler_class = self.scheduler_classes[0]
+        scheduler_config = self.get_scheduler_config()
+        scheduler = scheduler_class(**scheduler_config)
+
+        num_trained_timesteps = len(scheduler)
+
+        model = self.dummy_model()
```
Comment (anton-l) [inline tests/test_scheduler.py:870]:
```suggestion

        model = self.dummy_model()

        num_inference_steps = 10
        scheduler.set_timesteps(num_inference_steps)
```
This scheduler requires a `.set_timesteps()` call. Reference: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L225

File: tests/test_scheduler.py [hunk 3]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+        sample = self.dummy_sample_deter
```
Comment (anton-l) [inline tests/test_scheduler.py:871]:
```suggestion
        sample = self.dummy_sample_deter * scheduler.sigmas[0]
```
The initial sample needs to be multiplied by max sigma to make sure that it's within the supported range. Ref: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L229

File: tests/test_scheduler.py [hunk 4]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+
+        for t in reversed(range(num_trained_timesteps - 1)):
```
Comment (anton-l) [inline tests/test_scheduler.py:873]:
```suggestion
        for i, t in enumerate(scheduler.timesteps):
```

Can't get rid of enumeration yet, need both `i` and `t` for the loop in the current implementation :sweat_smile:

File: tests/test_scheduler.py [hunk 5]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+            # 1. predict noise residual
+            residual = model(sample, t)
+            # print("residual: ")
+            # print(residual)
```
Comment (anton-l) [inline tests/test_scheduler.py:877]:
```suggestion
            with torch.no_grad():
                sigma = scheduler.sigmas[i]
                sample = sample / ((sigma**2 + 1) ** 0.5)

                model_output = model(sample, t)
```

Need to rescale the model input here too, to conform to the ODE equation used in K-LMS. Ref: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L246

File: tests/test_scheduler.py [hunk 6]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+
+            # 2. predict previous mean of sample x_t-1
+            pred_prev_sample = scheduler.step(residual, t, sample).prev_sample
+
+            # if t > 0:
+            #     noise = self.dummy_sample_deter
+            #     variance = scheduler.get_variance(t) ** (0.5) * noise
+            #
+            # sample = pred_prev_sample + variance
+            sample = pred_prev_sample
```
Comment (anton-l) [inline tests/test_scheduler.py:887]:
```suggestion
            output = scheduler.step(model_output, i, sample)
            sample = output.prev_sample
```

The LMS sampler takes `i` instead of an explicit timestep

File: tests/test_scheduler.py [hunk 7]
```diff
@@ -822,3 +822,73 @@ def test_step_shape(self):
...
+        print("Result sample: ")
+        print(sample)
+        result_sum = torch.sum(torch.abs(sample))
+        result_mean = torch.mean(torch.abs(sample))
+
+        assert abs(result_sum.item() - 259.0883) < 1e-2
+        assert abs(result_mean.item() - 0.3374) < 1e-3
```
Comment (anton-l) [inline tests/test_scheduler.py:894]:
```suggestion
        assert abs(result_sum.item() - 1006.3885) < 1e-2
        assert abs(result_mean.item() - 1.3104) < 1e-3
```

These are the reference values if all of the above suggestions are applied, make sure they match on your hardware! (if not - we'll need to adjust)

**Review by anton-l — COMMENTED** [review]
Hi @sidthekidder thanks a lot for diving into these tests! The LMS scheduler differs quite a bit in its API from other schedulers (and needs refactoring, to be honest), so this is quite a challenge! 

I've left some suggestions to make the `full_loop` test work, let me know if something's not clear :) 
In general, since this scheduler is unique to Stable Diffusion at the moment, feel free to reference the SD pipeline for how it's used in practice.

File: tests/test_scheduler.py (hunk 1 above)
Comment (sidthekidder) [reply tests/test_scheduler.py]: Updated to add `scaled_linear` (`squaredcos_cap_v2` is not implemented for LMSDiscreteScheduler)

File: tests/test_scheduler.py (hunk 2 above)
Comment (sidthekidder) [reply tests/test_scheduler.py]: Added

File: tests/test_scheduler.py (hunk 3 above)
Comment (sidthekidder) [reply tests/test_scheduler.py]: Makes sense, updated

File: tests/test_scheduler.py (hunk 7 above)
Comment (sidthekidder) [reply tests/test_scheduler.py]: Adjusted values as necessary with the new output

**Comment by sidthekidder** [comment]
Thanks for the detailed comments @anton-l! I didn't realize we could follow the implementation in `pipeline_stable_diffusion.py`, thanks for linking to the references - helped me understand the flow better.

**Review by anton-l — COMMENTED** [review]
Thank you @sidthekidder! I think all that's left is to implement `test_pytorch_equal_numpy()` too :) 
And feel free to modify the scheduler if it fails this test for some reason!

**Comment by sidthekidder** [comment]
Added `test_pytorch_equal_numpy` similar to the `PNDMSchedulerTest.test_pytorch_equal_numpy` fn.

**Comment by patrickvonplaten** [comment]
Nice looks good to me :-)

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
@anton-l if you could review one more time this would be great :-)

File: tests/test_scheduler.py [hunk 8]
```diff
@@ -853,3 +853,89 @@ def test_step_shape(self):
...
+
+    def test_pytorch_equal_numpy(self):
+        for scheduler_class in self.scheduler_classes:
+            sample_pt = self.dummy_sample
+            residual_pt = 0.1 * sample_pt
+            dummy_past_residuals_pt = [residual_pt + 0.2, residual_pt + 0.15, residual_pt + 0.1, residual_pt + 0.05]
+
+            sample = sample_pt.numpy()
+            residual = 0.1 * sample
+            dummy_past_residuals = [residual + 0.2, residual + 0.15, residual + 0.1, residual + 0.05]
```
Comment (anton-l) [inline tests/test_scheduler.py:899]:
```suggestion
            sample_pt = self.dummy_sample
            sample = sample_pt.numpy()
```

File: tests/test_scheduler.py [hunk 9]
```diff
@@ -853,3 +853,89 @@ def test_step_shape(self):
...
+            scheduler_config["tensor_format"] = "pt"
+            scheduler_pt = scheduler_class(**scheduler_config)
+
+            scheduler.set_timesteps(self.num_inference_steps)
+            scheduler_pt.set_timesteps(self.num_inference_steps)
+
+            # copy over dummy past residuals (must be done after set_timesteps)
+            scheduler.ets = dummy_past_residuals[:]
+            scheduler_pt.ets = dummy_past_residuals_pt[:]
+
```
Comment (anton-l) [inline tests/test_scheduler.py:914]:
```suggestion
            scheduler_pt.set_timesteps(self.num_inference_steps)
            
```

**Review by anton-l — APPROVED** [review]
I'll just remove the unused past residuals (they're needed only in the PNDM tests) and we're good :)

**Comment by anton-l** [comment]
Thank you @sidthekidder!
